cdc/entry/schema/snapshot.go

```diff
@@ -0,0 +1,1151 @@
+	tables *btree.BTree
+
+	// map[versionedID] -> *model.TableInfo
+	partitions *btree.BTree
+
+	// map[versionedID] -> struct{}
+	truncatedTables *btree.BTree
+
+	// map[versionedID] -> struct{}
+	// Partitions and tables share ineligibleTables because their IDs won't conflict.
+	ineligibleTables *btree.BTree
+
+	// if forceReplicate is true, treat ineligible tables as eligible.
+	forceReplicate bool
+
+	currentTs uint64
+
+	rwlock *sync.RWMutex
+}
+
+// NewEmptySnapshot creates an empty schema snapshot.
+func NewEmptySnapshot(forceReplicate bool) *Snapshot {
+	return &Snapshot{
+		tableNameToID:    btree.New(16),
+		schemaNameToID:   btree.New(16),
+		schemas:          btree.New(16),
+		tables:           btree.New(16),
+		partitions:       btree.New(16),
+		truncatedTables:  btree.New(16),
+		ineligibleTables: btree.New(16),
+		forceReplicate:   forceReplicate,
+		currentTs:        0,
+		rwlock:           new(sync.RWMutex),
+	}
+}
+
+// NewSnapshotFromMeta creates a schema snapshot from meta.
+func NewSnapshotFromMeta(meta *timeta.Meta, currentTs uint64, forceReplicate bool) (*Snapshot, error) {
+	snap := NewEmptySnapshot(forceReplicate)
+	dbinfos, err := meta.ListDatabases()
+	if err != nil {
+		return nil, cerror.WrapError(cerror.ErrMetaListDatabases, err)
+	}
+	// `tag` is used to reverse sort all versions in the generated snapshot.
+	tag := negative(currentTs)
+
+	for _, dbinfo := range dbinfos {
+		vid := newVersionedID(dbinfo.ID, tag)
+		vid.target = unsafe.Pointer(dbinfo)
+		snap.schemas.ReplaceOrInsert(vid)
+
+		vname := newVersionedEntityName(-1, dbinfo.Name.O, tag) // -1 means the entity is a schema.
+		vname.target = dbinfo.ID
+		snap.schemaNameToID.ReplaceOrInsert(vname)
+
+		tableInfos, err := meta.ListTables(dbinfo.ID)
+		if err != nil {
+			return nil, cerror.WrapError(cerror.ErrMetaListDatabases, err)
+		}
+		for _, tableInfo := range tableInfos {
+			tableInfo := model.WrapTableInfo(dbinfo.ID, dbinfo.Name.O, currentTs, tableInfo)
+			snap.tables.ReplaceOrInsert(versionedID{
+				id:     tableInfo.ID,
+				tag:    tag,
+				target: unsafe.Pointer(tableInfo),
+			})
+			snap.tableNameToID.ReplaceOrInsert(versionedEntityName{
+				prefix: dbinfo.ID,
+				entity: tableInfo.Name.O,
+				tag:    tag,
+				target: tableInfo.ID,
+			})
+
+			ineligible := !tableInfo.IsEligible(forceReplicate)
+			if ineligible {
+				snap.ineligibleTables.ReplaceOrInsert(versionedID{id: tableInfo.ID, tag: tag})
+			}
+			if pi := tableInfo.GetPartitionInfo(); pi != nil {
+				for _, partition := range pi.Definitions {
+					vid := newVersionedID(partition.ID, tag)
+					vid.target = unsafe.Pointer(tableInfo)
+					snap.partitions.ReplaceOrInsert(vid)
+					if ineligible {
+						snap.ineligibleTables.ReplaceOrInsert(versionedID{id: partition.ID, tag: tag})
+					}
+				}
+			}
+		}
+	}
+
+	snap.currentTs = currentTs
+	return snap, nil
+}
+
+// Copy creates a new schema snapshot based on the given one.
+func (s *Snapshot) Copy() *Snapshot {
```

Comment:
This method seems to be shallow copying all the composite data structures and this seems to be the central idea that saves memory. Can you add more comments on the behavior of `Copy()`?